### ISSUE TITLE
OperatorsExpr parsing blows up on large file

### ISSUE THREAD

**Issue by AndrasKovacs** [agda/agda#5670]
On Agda 2.6.2, `Parsing.OperatorsExp` as reported by `-vprofile:7` takes 14 seconds on this [10k line file](https://github.com/AndrasKovacs/smalltt/blob/f9c27e825993d7e16282701d83941cf394f63725/bench/stlc10k.agda).

The file just repeats the same  ~130 lines containing a Church-encoded simply-typed LC, but all top-level symbols are uniquely renamed each time. The 5k line version of the same thing parses in 0,5 seconds, so there's probably something funny going on.

**Comment by andreasabel** [comment]
Should we maybe look at `flattenScope`?
https://github.com/agda/agda/blob/982e18cdcbb2355d571bc05a8a98811fe6b48525/src/full/Agda/Syntax/Concrete/Operators.hs#L182-L183 
Doing something which is linear in the scope for _every_ application might easily spoil the fun (quadratic!).

**Comment by nad** [comment]
Some other parts of the code might also be problematic:

https://github.com/agda/agda/blob/982e18cdcbb2355d571bc05a8a98811fe6b48525/src/full/Agda/Syntax/Concrete/Operators.hs#L184

https://github.com/agda/agda/blob/982e18cdcbb2355d571bc05a8a98811fe6b48525/src/full/Agda/Syntax/Concrete/Operators.hs#L278-L295

An operator grammar is built for each expression, and as an optimisation this is done based on the name parts in the expression. I guess it would be possible to maintain maps from name parts to names containing those name parts. One could still construct problematic examples, for instance if lots of operators contain the same name part, but I suspect that the problem that the code in the OP exhibits is more common in practice (note that this code does not contain any operators).

**Comment by andreasabel** [comment]
History:
- 73a28d600f383d5097cece7f42ddd018bff15db1 (Sep 2015 @nad):  Made (many) operator grammars smaller. This implies that some error messages become shorter.
- 703c9e9d44da01713b301911343c9a2414bce105 (Mar 2015 @nad): Made the operator parser roughly 60% faster (for the standard library, with my current setup) by making use of the knowledge of which name parts occur in the expression/pattern to be parsed.
- Qualified mixfix operators 3b4389931044b739539b8124a29cf2b7260b1479 (2011 @UlfNorell) 

One could try to backport the OP to Agda 2.3.2 and see whether there is the same performance degradation.

**Comment by nad** [comment]
The code is accepted by Agda 2.3.2 if every occurrence of the `OPTIONS` pragma except for the first one is removed. I got the following timings for 10 and 20 instances of the repeated code in the file:
```
  11,474,493,416 bytes allocated in the heap
   3,459,980,852 bytes copied during GC
     263,578,880 bytes maximum residency (24 sample(s))
       2,610,352 bytes maximum slop
             659 MB total memory in use (0 MB lost due to fragmentation)

                                    Tot time (elapsed)  Avg pause  Max pause
  Gen  0     21832 colls,     0 par    4.71s    4.76s     0.0002s    0.0040s
  Gen  1        24 colls,     0 par    2.93s    3.19s     0.1328s    0.6405s

  INIT    time    0.00s  (  0.00s elapsed)
  MUT     time    7.08s  (  7.15s elapsed)
  GC      time    7.63s  (  7.95s elapsed)
  EXIT    time    0.00s  (  0.04s elapsed)
  Total   time   14.72s  ( 15.13s elapsed)

  %GC     time      51.8%  (52.5% elapsed)

  Alloc rate    1,620,028,504 bytes per MUT second

  Productivity  48.1% of total user, 46.8% of total elapsed


real    0m15,138s
user    0m14,722s
sys     0m0,356s
```
```
  24,747,785,720 bytes allocated in the heap
   7,944,013,584 bytes copied during GC
     724,088,344 bytes maximum residency (24 sample(s))
       5,233,492 bytes maximum slop
            1727 MB total memory in use (0 MB lost due to fragmentation)

                                    Tot time (elapsed)  Avg pause  Max pause
  Gen  0     46755 colls,     0 par   10.49s   10.62s     0.0002s    0.0066s
  Gen  1        24 colls,     0 par    7.19s    7.89s     0.3289s    1.8593s

  INIT    time    0.00s  (  0.00s elapsed)
  MUT     time   16.63s  ( 16.76s elapsed)
  GC      time   17.68s  ( 18.52s elapsed)
  EXIT    time    0.01s  (  0.09s elapsed)
  Total   time   34.32s  ( 35.36s elapsed)

  %GC     time      51.5%  (52.4% elapsed)

  Alloc rate    1,487,924,888 bytes per MUT second

  Productivity  48.5% of total user, 47.1% of total elapsed


real    0m35,369s
user    0m34,325s
sys     0m0,864s
```
Agda 2.3.2 does not seem to give you any extra information if you use `-vprofile:7`.

**Comment by andreasabel** [comment]
> I got the following timings for 10 and 20 instances of the repeated code in the file:

How do these timings relate to the OP?

**Comment by nad** [comment]
What do you mean? I could add the information that Agda 2.6.3-2976b59 is, in both cases, a little faster.

**Comment by andreasabel** [comment]
> What do you mean? I could add the information that Agda 2.6.3-2976b59 is, in both cases, a little faster.

Yes, that is sufficient.  So we can assert that there is no regression, just the initial design does not scale.

**Comment by UlfNorell** [comment]
I did some experiments in #5670.

The underlying problem is that operator parsing needs to account for everything that's in scope. For instance

```agda
foo _op_ x y = x op y
```

parses as `_op_ x y` unless there's already an `op` in scope, in which case it's ambiguous (the local `_op_` does not shadow an outer `op`).

To do the parsing we basically need to know the names in scope (qualified or otherwise) and their fixities. We compute these using `flattenScope` and `localNames` in `Agda.Syntax.Scope.Flat`. This happens for every expression we parse and it's linear in the current scope, causing the quadratic behaviour in the OP. In the linked PR I tried to cache this information and only recompute it when the scope changes, and this shaved of about 50% of the time, but caused memory issues in the standard library.

I think the right solution is along the lines of what @nad suggested: driving the parser construction entirely from the name parts used in the expression, and having a reasonably efficient way to look up names from name parts.

**Comment by andreasabel** [comment]
Something for the Hackathon, maybe:

> having a reasonably efficient way to look up names from name parts.

**Comment by AndrasKovacs** [comment]
Fixed by #8079
